### PR TITLE
Sign apk for Android release builds

### PR DESCRIPTION
**Pull request by samartnik**
## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions.

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by samartnik** [comment]
Closes https://github.com/brave/brave-browser/issues/7383

File: lib/util.js [hunk 1]
```diff
@@ -367,6 +367,20 @@ const util = {
 
   signApp: (options = config.defaultOptions) => {
     console.log('signing ...')
+    if (config.targetOS === 'android') {
+      const apkDir = path.join(config.outputDir, 'apks')
+      const fullApkPath = path.join(apkDir, config.targetApkName)
+      const fullApkPathAligned = fullApkPath + '_aligned'
+      util.run('third_party/android_sdk/public/build-tools/29.0.2/zipalign', ['-f', '-p', 4, fullApkPath, fullApkPathAligned], options)
```
Comment (bridiver) [inline lib/util.js]: So we're actually trying specifically to keep this kind of stuff out of js. Can you create a `sign_app` target in brave/build/android instead? Also it seems like the the `android_apk` template already does this through`package_apk`?

Comment (bridiver) [reply lib/util.js]:
```  # Creates a signed and aligned .apk.
  #
  # Variables
  #   apk_name: (optional) APK name (without .apk suffix). If provided, will
  #       be used to generate .info files later used by the supersize tool.
  #   assets_build_config: Path to android_apk .build_config containing merged
  #       asset information.
  #   deps: Specifies the dependencies of this target.
  #   dex_path: Path to classes.dex file to include (optional).
  #   packaged_resources_path: Path to .ap_ to use.
  #   output_apk_path: Output path for the generated .apk.
  #   native_lib_placeholders: List of placeholder filenames to add to the apk
  #     (optional).
  #   secondary_native_lib_placeholders: List of placeholder filenames to add to
  #     the apk for the secondary ABI (optional).
  #   native_libs: List of native libraries.
  #   native_libs_filearg: @FileArg() of additionally native libraries.
  #   secondary_abi_native_libs: (optional) List of native libraries for
  #     secondary ABI.
  #   secondary_abi_native_libs_filearg: (optional). @FileArg() of additional
  #     secondary ABI native libs.
  #   write_asset_list: Adds an extra file to the assets, which contains a list of
  #     all other asset files.
  #   keystore_path: Path to keystore to use for signing.
  #   keystore_name: Key alias to use.
  #   keystore_password: Keystore password.
  #   uncompress_shared_libraries: (optional, default false) Whether to store
  #     native libraries inside the APK uncompressed and page-aligned.```

Comment (samartnik) [reply lib/util.js]: Using `package_apk` would be great, but for some reason they omitted private key password (`--key-pass`) parameter https://chromium.googlesource.com/chromium/src.git/+/refs/tags/79.0.3945.92/build/android/gyp/finalize_apk.py#27
We won't be able to use it without patching Chromium files.
So I'll make it through `sign_app` target as it doesn't require patching files.

Comment (bridiver) [reply lib/util.js]:
```
"--key-passwd",
invoker.keystore_password,
``` ?

Comment (samartnik) [reply lib/util.js]: it is `--ks-pass`, we also need `--key-pass`. We need 2 passwords.

**Comment by samartnik** [comment]
The follow up issue to use `package_apk` instead https://github.com/brave/brave-browser/issues/7404

File: lib/config.js [hunk 2]
```diff
@@ -129,6 +133,10 @@ Config.prototype.buildArgs = function () {
     brave_referrals_api_key: this.braveReferralsApiKey,
     enable_hangout_services_extension: this.enable_hangout_services_extension,
     enable_cdm_host_verification: this.brave_enable_cdm_host_verification,
+    brave_android_keystore_path: this.androidKeystorePath,
```
Comment (bridiver) [inline lib/config.js]: see https://github.com/brave/brave-core/pull/4233/files#r358508527

Comment (bridiver) [reply lib/config.js]: Also please make sure all the names are consistent. If we have `brave_android_key_password` it should map to `this.braveAndroidKeyPassword` and `getNPMConfig(['brave_android_key_password'])`

File: lib/config.js [hunk 3]
```diff
@@ -250,10 +258,12 @@ Config.prototype.buildArgs = function () {
 Config.prototype.shouldSign = function () {
   // it doesn't make sense to sign debug builds because the restrictions on loading
   // dynamic libs prevents them from working anyway
-  return this.mac_signing_identifier !== '' &&
-         !this.skip_signing &&
+  return !this.skip_signing &&
          this.buildConfig === 'Release' &&
-         this.targetOS !== 'ios'
+         (this.mac_signing_identifier !== '' &&
```
Comment (bridiver) [inline lib/config.js]: can you add nested `()` here for ios vs android with the `||`?

Comment (bridiver) [reply lib/config.js]: I know it works fine because of operator precedence, but it helps readability

File: lib/config.js [hunk 4]
```diff
@@ -83,6 +83,10 @@ const Config = function () {
   this.extraGnArgs = {}
   this.extraNinjaOpts = []
   this.braveSafetyNetApiKey = getNPMConfig(['brave_safetynet_api_key']) || ''
+  this.androidKeystorePath = getNPMConfig(['android_keystore_path']) || ''
+  this.androidKeyAlias = getNPMConfig(['android_key_alias']) || ''
```
Comment (bridiver) [inline lib/config.js]: shouldn't this be `android_keystore_name`?

Comment (samartnik) [reply lib/config.js]: In apksigner this param called `--ks-key-alias`. I don't really care how to name. Will change this.

**Comment by bridiver** [comment]
Is it? I didn’t see that param in the upstream script

File: lib/config.js (hunk 2 above)
Comment (samartnik) [reply lib/config.js]: Changed the names of variables

File: lib/config.js (hunk 3 above)
Comment (samartnik) [reply lib/config.js]: Done

File: lib/config.js (hunk 4 above)
Comment (samartnik) [reply lib/config.js]: Changed the names of variables

**Pull request merged**